lib/srv/reexec.go

```diff
@@ -1194,10 +1202,42 @@ func copyCommand(ctx *ServerContext, cmdmsg *ExecCommand) {
 	}
 }
 
-// CheckHomeDir checks if the user's home dir exists
+// HasAccessibleHomeDir checks if the given User has access to an existing home directory.
+func HasAccessibleHomeDir(localUser *user.User) error {
+	fi, err := os.Stat(localUser.HomeDir)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if !fi.IsDir() {
+		return trace.NotFound("user home is not a directory")
+	}
+
+	stat, ok := fi.Sys().(*syscall.Stat_t)
+	if !ok {
+		return trace.AccessDenied("could not retrieve owner of home directory")
+	}
+
+	uid := strconv.Itoa(int(stat.Uid))
+	gid := strconv.Itoa(int(stat.Gid))
+
+	if uid == localUser.Uid && gid == localUser.Gid {
+		return nil
+	}
+
+	return trace.AccessDenied("user does not own configured home directory")
+}
+
+// CheckHomeDir checks if the user's home dir exists and is accessible. This also handles cases where
+// the home directory isn't visible to the root user, which HasAccessibleHomeDir doesn't account for.
 func CheckHomeDir(localUser *user.User) (bool, error) {
-	if fi, err := os.Stat(localUser.HomeDir); err == nil {
-		return fi.IsDir(), nil
+	err := HasAccessibleHomeDir(localUser)
```

Comment:
Funnily enough, simple `os.Stat()` is likely more complete check due to leveraging the actual system stack to do the check.

What is the problem with the root user?

Reply:
> What is the problem with the root user?

Can you add to this? I'm not sure I understand what you're asking 🤔